physical/redis.go

```diff
@@ -0,0 +1,314 @@
+	realPrefix := fmt.Sprintf("%s/%s", c.path, prefix)
+	if !strings.HasSuffix(realPrefix, "/") {
+		realPrefix += "/"
+	}
+
+	// Ask Redis to list all keys beneath our prefix.  Note that this
+	// is not terribly efficient if you have a lot of keys, but we hope
+	// it's not a common operating.  There are more complex APIs for
+	// doing this incrementally.
+	reply, err := conn.Do("KEYS", realPrefix+"*")
+	matches, err := redis.Strings(reply, err)
+
+	// TODO - Don't recurse subdirectories.  This means stripping
+	// everything after "/" and removing duplicates.
+	results := make([]string, 0)
+	dirs := make(map[string]bool)
+	for _, match := range matches {
+		// Remove the path that we were querying from the results.
+		match = strings.TrimPrefix(match, realPrefix)
+
+		slashPos := strings.Index(match, "/")
+		if slashPos == -1 {
+			// We have an ordinary file, so return it directly.
+			results = append(results, match)
+		} else {
+			// Keep the trailing slash.
+			dir := match[0 : slashPos+1]
+
+			// Have we seen this directory already?  If not,
+			// return it and mark it as seen.
+			if _, present := dirs[dir]; !present {
+				results = append(results, dir)
+				dirs[dir] = true
+			}
+		}
+	}
+
+	return results, err
+}
+
+func (c *RedisBackend) LockWith(key, value string) (Lock, error) {
+	return &RedisLock{
+		key:    fmt.Sprintf("%s/_lock/%s", c.path, key),
+		value:  value,
+		pool:   c.pool,
+		logger: c.logger,
+	}, nil
+}
+
+type RedisLock struct {
+	key    string
+	value  string
+	pool   *redis.Pool
+	logger *log.Logger
+}
+
+// Acquire the lock.  To interrupt lock acquistion, close stopCh.  The
+// returned channel will be closed if the lock is lost.
+func (c *RedisLock) Lock(stopCh <-chan struct{}) (<-chan struct{}, error) {
+	// Loop until we get a non-nil reply from Redis indicating a
+	// successful operation.
+	var reply interface{}
+	for reply == nil {
+		conn := c.pool.Get()
+		defer conn.Close()
+
+		// Attempt to set our key.  "NX" means to only set the key
+		// if it does not exist, and "PX" specifies a timeout in
+		// milliseconds.
+		reply, err := conn.Do("SET", c.key, c.value, "NX", "PX", "30000")
+		if err != nil {
+			// We got an error communicating with Redis, so
+			// fail outright.
+			return nil, err
+		}
+		if reply != nil {
+			// We got back a non-nil response, which means we
+			// set the key and can stop looping.
+			break
+		}
+
+		// Wait a while before retrying.
+		select {
+		case <-stopCh:
+			// Lock acquisition was cancelled by our caller.
+			return nil, nil
+		case <-time.After(5 * time.Second):
+			// Timeout fired, so here we go again.
+		}
+	}
+
+	// Set up a background listener to renew our lock and notice if it
+	// goes away.
+	leaderCh := make(chan struct{})
+	go func() {
+		conn := c.pool.Get()
+		defer conn.Close()
+
+		// Create our TTL-bumping script and try to load
+		// it. Loading is optional because of how Redigo is
+		// implemented.
+		bumpTtlScript := redis.NewScript(1, `
```

Comment:
technically you could define this function when you connect to redis for the first time and then refer to it by its sha

instead of constantly redefining